### PR TITLE
testing: raise bottom-up test from snitch_stream to memref_stream

### PR DESCRIPTION
**Pull request by superlopuh**
Once this PR is merged, we know that a representation of the kernels using memref_stream can generate assembly we believe to be optimal for Snitch. While raising the level of abstraction of the inputs, I found a number of missing features in the compiler.

- [x] testing: manually optimise snitch_stream stride patterns in bottom-up test
[#2267](https://github.com/xdslproject/xdsl/issues/2267)
- [x] dialects: (snitch_stream) simplify bounds of size 1 in stride pattern
[#2269](https://github.com/xdslproject/xdsl/issues/2269)
- [x] transformations: add imperfect loop nest support in memref_stream.streaming_region
[#2268](https://github.com/xdslproject/xdsl/issues/2268)

### CODE REVIEW

File: tests/filecheck/projects/riscv-backend-paper/bottom_up.mlir [hunk 1]
```diff
@@ -57,14 +60,14 @@ func.func public @conv_2d_nchw_fchw_d1_s1_3x3(
 // CHECK-NEXT:      mv t2, a1
 // CHECK-NEXT:      mv t0, a2
 // CHECK-NEXT:      li t5, 8
+// CHECK-NEXT:      li a5, 2
```
Comment (superlopuh) [inline tests/filecheck/projects/riscv-backend-paper/bottom_up.mlir:63]: #2267

File: tests/filecheck/projects/riscv-backend-paper/bottom_up.mlir [hunk 2]
```diff
@@ -186,35 +185,38 @@ func.func public @conv_2d_nchw_fchw_d1_s1_3x3(
...
-      ],
-      "operandSegmentSizes" = array<i32: 3, 0>
-    }> ({
-    ^bb0(%X_stream : !stream.readable<!riscv.freg<ft0>>, %W_stream : !stream.readable<!riscv.freg<ft1>>, %B_stream : !stream.readable<!riscv.freg<ft2>>):
+    memref_stream.streaming_region {
+      bounds = [8, 8, 8],
+      indexing_maps = [
+          affine_map<(m, n, k) -> (m, k)>,
+          affine_map<(m, n, k) -> (k, n)>,
+          affine_map<(m, n) -> (m, n)>
```
Comment (superlopuh) [inline tests/filecheck/projects/riscv-backend-paper/bottom_up.mlir:193]: #2268

File: xdsl/dialects/snitch_stream.py [hunk 3]
```diff
@@ -152,25 +152,31 @@ def simplified(self) -> StridePattern:
         """
         if len(self.ub) < 2:
             return self
-        d = -2
+
+        tuples = tuple(
```
Comment (superlopuh) [inline xdsl/dialects/snitch_stream.py:156]: #2269

File: xdsl/transforms/convert_memref_stream_to_snitch_stream.py [hunk 4]
```diff
@@ -110,7 +110,7 @@ def match_and_rewrite(
         # TODO: make consistent
         stride_patterns = tuple(
             snitch_stream.StridePattern.from_bounds_and_strides(
-                bounds[::-1], strides[::-1]
+                bounds[: len(strides)][::-1], strides[::-1]
```
Comment (superlopuh) [inline xdsl/transforms/convert_memref_stream_to_snitch_stream.py:113]: #2268

File: tests/filecheck/projects/riscv-backend-paper/bottom_up.mlir [hunk 5]
```diff
@@ -1,4 +1,4 @@
-// RUN: xdsl-opt -p convert-func-to-riscv-func,reconcile-unrealized-casts,test-lower-snitch-stream-to-asm -t riscv-asm %s | filecheck %s
+// RUN: xdsl-opt -p convert-func-to-riscv-func,convert-memref-stream-to-snitch,reconcile-unrealized-casts,test-lower-snitch-stream-to-asm -t riscv-asm %s | filecheck %s
```
Comment (superlopuh) [inline tests/filecheck/projects/riscv-backend-paper/bottom_up.mlir:1]: note that the assembly generated is exactly the same as before

**Review by compor — APPROVED** [review]

**Pull request merged**